File: android/src/main/kotlin/dev/aditya/apps_handler/AppInfo.kt
package dev.aditya.apps_handler

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.os.Build
import java.io.ByteArrayOutputStream

data class AppInfo(
    val name: String,
    val packageName: String,
    val category: String,
    val versionName: String?,
    val versionCode: Long,
    val dataDir: String,
    val systemApp: Boolean,
    val installerPackageName: String?,
    private val drawable: Drawable?,
    val enabled: Boolean,
    val installTimeMillis: Long,
    val updateTimeMillis: Long,
) {
    companion object {
        fun createFromPackageInfo(
            context: Context,
            packageInfo: PackageInfo,
            loadIcon: Boolean = false
        ): AppInfo? {
            return try {
                val packageManager = context.packageManager
                val applicationInfo = packageInfo.applicationInfo ?: return null
                val category = getApplicationCategory(applicationInfo)

                AppInfo(
                    name = applicationInfo.loadLabel(packageManager).toString(),
                    packageName = packageInfo.packageName,
                    category = category,
                    versionName = packageInfo.versionName,
                    versionCode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                        packageInfo.longVersionCode
                    } else {
                        @Suppress("DEPRECATION")
                        packageInfo.versionCode.toLong()
                    },
                    dataDir = applicationInfo.dataDir,
                    systemApp = (applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM) != 0,
                    installerPackageName = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                        packageManager.getInstallSourceInfo(packageInfo.packageName).installingPackageName
                    } else {
                        @Suppress("DEPRECATION")
                        packageManager.getInstallerPackageName(packageInfo.packageName)
                    },
                    drawable = if (loadIcon) applicationInfo.loadIcon(packageManager) else null,
                    enabled = applicationInfo.enabled,
                    installTimeMillis = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                        packageInfo.firstInstallTime
                    } else {
                        @Suppress("DEPRECATION")
                        packageInfo.firstInstallTime
                    },
                    updateTimeMillis = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                        packageInfo.lastUpdateTime
                    } else {
                        @Suppress("DEPRECATION")
                        packageInfo.lastUpdateTime
                    }
                )
            } catch (e: Exception) {
                null
            }
        }

        private fun getApplicationCategory(applicationInfo: ApplicationInfo): String {
            return when {
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> {
                    when (applicationInfo.category) {
                        ApplicationInfo.CATEGORY_GAME -> "game"
                        ApplicationInfo.CATEGORY_AUDIO -> "audio"
                        ApplicationInfo.CATEGORY_VIDEO -> "video"
                        ApplicationInfo.CATEGORY_IMAGE -> "image"
                        ApplicationInfo.CATEGORY_SOCIAL -> "social"
                        ApplicationInfo.CATEGORY_NEWS -> "news"
                        ApplicationInfo.CATEGORY_MAPS -> "maps"
                        ApplicationInfo.CATEGORY_PRODUCTIVITY -> "productivity"
                        else -> "undefined"
                    }
                }
                else -> "undefined"
            }
        }
    }

    fun toMap(includeIcon: Boolean = false): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>( 
            "app_name" to name,
            "package_name" to packageName,
            "category" to category,
            "version_name" to versionName,
            "version_code" to versionCode,
            "data_dir" to dataDir,
            "system_app" to systemApp,
            "installer_package_name" to installerPackageName,
            "enabled" to enabled,
            "install_time" to installTimeMillis,
            "update_time" to updateTimeMillis
        )

        if (includeIcon && drawable != null) {
            map["app_icon"] = drawableToByteArray(drawable)
        }

        return map
    }

    private fun drawableToByteArray(drawable: Drawable): ByteArray {
        val bitmap = Bitmap.createBitmap(
            drawable.intrinsicWidth,
            drawable.intrinsicHeight,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)

        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
        val byteArray = stream.toByteArray()
        bitmap.recycle()
        return byteArray
    }
}

File: lib/src/models/app_info.dart
class AppInfo {
  final String appName;
  final String packageName;
  final String category;
  final String? versionName;
  final int versionCode;
  final String dataDir;
  final bool systemApp;
  final String? installerPackageName;
  final bool enabled;
  final int installTime;
  final int updateTime;
  final List<int>? appIcon;

  const AppInfo({
    required this.appName,
    required this.packageName,
    required this.category,
    required this.versionName,
    required this.versionCode,
    required this.dataDir,
    required this.systemApp,
    required this.installerPackageName,
    required this.enabled,
    required this.installTime,
    required this.updateTime,
    this.appIcon,
  });

  factory AppInfo.fromMap(Map<dynamic, dynamic> map) {
    return AppInfo(
      appName: map['app_name'] as String,
      packageName: map['package_name'] as String,
      category: map['category'] as String,
      versionName: map['version_name'] as String?,
      versionCode: map['version_code'] as int,
      dataDir: map['data_dir'] as String,
      systemApp: map['system_app'] as bool,
      installerPackageName: map['installer_package_name'] as String?,
      enabled: map['enabled'] as bool,
      installTime: map['install_time'] as int,
      updateTime: map['update_time'] as int,
      appIcon: map['app_icon'] != null ? List<int>.from(map['app_icon']) : null,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'app_name': appName,
      'package_name': packageName,
      'category': category,
      'version_name': versionName,
      'version_code': versionCode,
      'data_dir': dataDir,
      'system_app': systemApp,
      'installer_package_name': installerPackageName,
      'enabled': enabled,
      'install_time': installTime,
      'update_time': updateTime,
      'app_icon': appIcon,
    };
  }

  @override
  String toString() {
    return 'AppInfo(appName: $appName, packageName: $packageName, category: $category, '
        'versionName: $versionName, versionCode: $versionCode, dataDir: $dataDir, '
        'systemApp: $systemApp, installerPackageName: $installerPackageName, '
        'enabled: $enabled, installTime: $installTime, updateTime: $updateTime, '
        'hasIcon: ${appIcon != null})';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is AppInfo && other.packageName == packageName;
  }

  @override
  int get hashCode => packageName.hashCode;
}
